packages/api-server/rest_server/tasks/test_dispatcher.py

```diff
@@ -0,0 +1,135 @@
+import unittest
+from unittest.mock import MagicMock
+
+import rclpy
+from rclpy.node import Client, Node
+
+from .dispatcher import DispatcherClient
+
+Client.wait_for_service = MagicMock(return_value=True)
+rclpy.init(args=None)
+dispatcher_client = DispatcherClient()
+
+
+class TestDispatcherClient(unittest.TestCase):
+    def test_convert_task_request(self):
+        mock_task_json_1 = {"task_type": "Loop", "start_time": 0}
+        req_msg_1, err_msg_1 = dispatcher_client.convert_task_request(mock_task_json_1)
+        self.assertEqual(err_msg_1, "Key value is incomplete")
+        self.assertEqual(req_msg_1, None)
+
+        mock_task_json_2 = {"task_type": "Loop", "description": {}}
+        req_msg_2, err_msg_2 = dispatcher_client.convert_task_request(mock_task_json_2)
+        self.assertEqual(err_msg_2, "Key value is incomplete")
+        self.assertEqual(req_msg_2, None)
+
+        mock_task_json_3 = {"start_time": 0, "description": {}}
+        req_msg_3, err_msg_3 = dispatcher_client.convert_task_request(mock_task_json_3)
+        self.assertEqual(err_msg_3, "Key value is incomplete")
+        self.assertEqual(req_msg_3, None)
+
+        mock_task_json_4 = {
+            "task_type": "Loop",
+            "start_time": 0,
+            "description": {},
+            "priority": "-1",
+        }
+        req_msg_4, err_msg_4 = dispatcher_client.convert_task_request(mock_task_json_4)
+        self.assertEqual(err_msg_4, "Priority value is less than 0")
+        self.assertEqual(req_msg_4, None)
+
+        mock_task_json_5 = {
+            "task_type": "Clean",
+            "start_time": 0,
+            "description": {"cleaning_zone": "zone_2"},
+            "priority": "0",
+        }
+        req_msg_5, err_msg_5 = dispatcher_client.convert_task_request(mock_task_json_5)
+        self.assertEqual(err_msg_5, "")
+        self.assertIsNotNone(req_msg_5)
+
+        mock_task_json_6 = {
+            "task_type": "Loop",
+            "start_time": 0,
+            "description": {
+                "num_loops": 1,
+                "start_name": "start",
+                "finish_name": "finish",
+            },
+        }
+        req_msg_6, err_msg_6 = dispatcher_client.convert_task_request(mock_task_json_6)
+        self.assertEqual(err_msg_6, "")
+        self.assertIsNotNone(req_msg_6)
+
+        mock_task_json_7 = {
+            "task_type": "Delivery",
+            "start_time": 0,
+            "description": {
+                "option": "coke",
+                "pickup_place_name": "coe",
+                "pickup_dispenser": "coke_dispenser",
+                "dropoff_ingestor": "coke_ingestor",
+                "dropoff_place_name": "supplies",
+            },
+        }
+        req_msg_7, err_msg_7 = dispatcher_client.convert_task_request(mock_task_json_7)
+        self.assertEqual(err_msg_7, "")
+        self.assertIsNotNone(req_msg_7)
+
+        mock_task_json_8 = {
+            "task_type": "wrong_task",
```

Comment:
@Yadunund Who should be verifying the input? Should dispatcher node response with a error message if it can't process the task?

Reply:
The result of the service call to the dispatcher will be [unsuccessful](https://github.com/open-rmf/rmf_internal_msgs/blob/b6a583250948dcf6fd32fe32a1d8ba6a3d7a3787/rmf_task_msgs/srv/SubmitTask.srv#L18) if a task type that is not enumerated in [TaskType.msg](https://github.com/open-rmf/rmf_internal_msgs/blob/b6a583250948dcf6fd32fe32a1d8ba6a3d7a3787/rmf_task_msgs/msg/TaskType.msg#L2-L7) is received. But the `unsuccessful` result will also be received in situations where the request fields are incomplete. So it may be worthwhile for the UI to verify the input as well so that the user can be informed the exact reason the task cannot be processed (incomplete field, invalid task type, etc)

Reply:
Can the dispatcher node give the error message as well? It's going to be very hard for the api server to verify more complex inputs. e.g. Checking start and end actually exists, checking certain field is required for certain task types. This would eventually lead to problems like the UI not allowing to submit tasks that should be possible (because new updates in dispatcher node has new rules), or the UI allowing tasks the should not be possible to be submitted, and the dispatcher node assumes that the input are verified correctly when it is not, leading to undefined behaviors.

Reply:
There is also the UX problem that, for rules the UI doesn't/cannot check, and the dispatcher node returns with an error, all it can do is tell the user an unknown error occured. Since the dispatcher node already does the checking, it would really help if it can return an error message so the user will know what is wrong.

Reply:
That is a fair point. @tanyouliang95 , do you think we can add a `string` field to the response of `SubmitTask.srv` that contains an error message when the request is unsuccessful?

Reply:
Sure, we can add an additional field for this. Similarly, this will also apply to `CancelTask.srv`. Tbh, currently the possible error during submit task srv can only be related to incorrect or incomplete field in `TaskDescription`. More variation of Error will occur after submission.

On top of the verbose string, one future proposal is to add a new msg in  `rmf_task_msgs`, which could be named as `TaskErrorCode.msg`. This will consist of an enumeration of possible task-related error in RMF, which can also part of `TaskStatus.msg` for verbose informing of error.